azure-maven-plugin-lib/src/main/java/com/microsoft/azure/maven/AbstractAzureMojo.java

```diff
@@ -272,72 +290,117 @@ public int getHttpProxyPort() {
+                    TextUtils.blue(SubscriptionOption.getSubscriptionName(subscriptions[0]))));
+            return subscriptions[0].subscriptionId();
+        }
+        final String defaultId = Optional.ofNullable(az.getCurrentSubscription()).map(Subscription::subscriptionId).orElse(null);
+        final List<SubscriptionOption> wrapSubs = Arrays.stream(subscriptions).map(t -> new SubscriptionOption(t))
+                .sorted()
+                .collect(Collectors.toList());
+        final SubscriptionOption defaultValue = wrapSubs.stream()
+                .filter(t -> StringUtils.equalsIgnoreCase(t.getSubscriptionId(), defaultId)).findFirst().orElse(null);
+        final TextIO textIO = TextIoFactory.getTextIO();
+        final SubscriptionOption subscriptionOptionSelected = new CustomTextIoStringListReader<SubscriptionOption>(() -> textIO.getTextTerminal(), null)
+                .withCustomPrompt(String.format("Please choose a subscription%s: ",
+                        highlightDefaultValue(defaultValue == null ? null : defaultValue.getSubscriptionName())))
+                .withNumberedPossibleValues(wrapSubs).withDefaultValue(defaultValue).read("Available subscriptions:");
+        if (subscriptionOptionSelected == null) {
+            throw new AzureExecutionException("You must select a subscription.");
+        }
+        return subscriptionOptionSelected.getSubscription().subscriptionId();
+    }
+
+    protected Azure createAzureClient() throws AzureAuthFailureException, AzureExecutionException {
         try {
-            final AzureEnvironment environment;
-            if (!AzureAuthHelper.validateEnvironment(auth.getEnvironment())) {
-                Log.prompt(String.format(UNSUPPORTED_AZURE_ENVIRONMENT, auth.getEnvironment()));
-                environment = AzureEnvironment.AZURE;
-            } else {
-                environment = AzureAuthHelper.getAzureEnvironment(auth.getEnvironment());
+            MavenAuthConfiguration authConfiguration = auth;
+            if (authConfiguration == null) {
+                authConfiguration = new MavenAuthConfiguration();
+            }
+            authConfiguration.setType(getAuthType());
+            final String serverId = authConfiguration.getServerId();
+            if (StringUtils.isNotBlank(serverId)) {
+                authConfiguration = MavenSettingHelper.buildAuthConfigurationByServerId(session, settingsDecrypter, serverId);
+            }
+            try {
+                validateAuthConfiguration(authConfiguration);
+            } catch (InvalidConfigurationException ex) {
+                final String messagePostfix = StringUtils.isNotBlank(serverId) ? ("in server: '" + serverId + "' at maven settings.xml.")
+                        : "in <auth> configuration.";
+                throw new AzureExecutionException(String.format("%s %s", ex.getMessage(), messagePostfix));
+            }
```

Comment:
this method seems too long, how about extracting into smaller methods: e.g. the upper part is preparing auth configuration. so
⬆️ `initAuthConfiguration`